### DISCUSSION TITLE
JSON error responses should have relevant HTTP status codes

### DISCUSSION THREAD

**Discussion by timkelty** [craftcms/cms#8631] [Ideas] [answered]
Throughout Craft's controllers, anywhere that returns JSON with errors, the response is a 200.

It really should be 4xx or 5xx, so it can be properly handled.

A good solution might be to add this to `craft\web\Controller::asErrorJson`, and use that more universally.

I'm thinking something like this might be ideal:

```
public function asErrorJson($data, int $statusCode = 500): YiiResponse
{
    $data = is_string($data) ? ['error' => $error] : $data;
    
    return $this->asJson($data)
        ->setStatusCode($statusCode);
}
```

Or perhaps something else to unify all JSON responses (success, error, errors, message, etc: https://github.com/craftcms/cms/issues/2495)

**Comment by brandonkelly** [comment]
Agree this would be better. However it would be a breaking change for anyone creating these requests using jQuery Ajax methods, as the `success` method would no longer be called on non-200 responses. (An improvement to be sure but will have to wait until Craft 4.)

**Comment by timkelty** [comment]
Thanks @brandonkelly.

In the meantime I wonder if I can duck-punch `craft\web\Controller` with an override of `asJson` that checks for `error(s)`  in the params – and load it in `config/app.php`?

**Comment by brandonkelly** [comment]
You could hack it in by creating a module with this:

```php
use yii\base\Event;
use yii\web\Response;

Event::on(Response::class, Response::EVENT_BEFORE_SEND, function(Event $e) {
    /** @var Response $response */
    $response = $e->sender;

    if (
        $response->format === Response::FORMAT_JSON &&
        !empty($response->data['error']) &&
        $response->getStatusCode() === 200 &&
        \Craft::$app->getRequest()->getIsSiteRequest()
    ) {
        $response->setStatusCode(400);
    }
});
```

**Comment by timkelty** [comment, marked as answer]
Hey @timkelty, it's you from the future!
We did it, buddy!

https://github.com/craftcms/cms/pull/10605

**Comment by TomDeSmet** [comment]
I just stumbled upon this and I actually disagree with this change. If you look at the docs it clearly says:

> The HyperText Transfer Protocol (HTTP) 400 Bad Request response status code indicates that the server cannot or will not process the request **due to something that is perceived to be a client error** (for example, malformed request syntax, invalid request message framing, or deceptive request routing).

Doing something like `updateCart()` while the stock of the product is empty now returns a Bad Request error. This is not a bad request, the client part did not fail here. The request reached the server, the server processed my request but halted due to a condition on **the server** not meeting expectations.

I feel this implementation, as a breaking change, is not a step forward. When I get a 400 I expect my client code to be wrong, my posted data to be wrong (e.g. bad CSRF), but not that the server did not complete my request.

**Reply by timkelty** [reply]
@TomDeSmet When it comes to HTTP code semantics, there is often a lot of room for interpretation.

You may be right that there may be a more appropriate code for something like out of stock, but as far as this change was concerned, the most important thing is that the code is 400 or up, meaning that it will be considered an error by anything processing the request.

Most applicably, `fetch()` requests will reject the Promise so it can be handled in the `.catch()`.
